sim/firesim-lib/src/main/cc/bridges/tracerv.cc

```diff
@@ -232,64 +228,100 @@ size_t tracerv_t::process_tokens(int num_beats, int minimum_batch_beats) {
+        fprintf(tracefile, "%016lx", OUTBUF[i + 6]);
+        fprintf(tracefile, "%016lx", OUTBUF[i + 5]);
+        fprintf(tracefile, "%016lx", OUTBUF[i + 4]);
+        fprintf(tracefile, "%016lx", OUTBUF[i + 3]);
+        fprintf(tracefile, "%016lx", OUTBUF[i + 2]);
+        fprintf(tracefile, "%016lx", OUTBUF[i + 1]);
+        fprintf(tracefile, "%016lx\n", OUTBUF[i + 0]);
+        // At least one valid instruction
+      } else {
+        for (int q = 0; q < max_consider; q++) {
+          if (OUTBUF[i + q + 1] & valid_mask) {
+            fprintf(tracefile,
+                    "Cycle: %016" PRId64 " I%d: %016" PRIx64 "\n",
+                    OUTBUF[i + 0],
+                    q,
+                    OUTBUF[i + q + 1] & (~valid_mask));
+          } else {
+            break;
           }
         }
       }
-    } else if (this->fireperf) {
+    }
+  } else if (fireperf) {
 
-      for (int i = 0; i < (bytes_received / sizeof(uint64_t)); i += 8) {
-        uint64_t cycle_internal = OUTBUF[i + 0];
+    for (int i = 0; i < (bytes_received / sizeof(uint64_t)); i += 8) {
+      uint64_t cycle_internal = OUTBUF[i + 0];
 
-        for (int q = 0; q < max_core_ipc; q++) {
-          if (OUTBUF[i + 1 + q] & valid_mask) {
-            uint64_t iaddr =
-                (uint64_t)((((int64_t)(OUTBUF[i + 1 + q])) << 24) >> 24);
-            this->trace_tracker->addInstruction(iaddr, cycle_internal);
+      for (int q = 0; q < max_consider; q++) {
+        if (OUTBUF[i + 1 + q] & valid_mask) {
+          uint64_t iaddr =
+              (uint64_t)((((int64_t)(OUTBUF[i + 1 + q])) << 24) >> 24);
+          addInstruction(iaddr, cycle_internal);
 #ifdef FIREPERF_LOGGER
-            fprintf(this->tracefile, "%016llx", iaddr);
-            fprintf(this->tracefile, "%016llx\n", cycle_internal);
+          fprintf(tracefile, "%016llx", iaddr);
+          fprintf(tracefile, "%016llx\n", cycle_internal);
 #endif // FIREPERF_LOGGER
-          }
         }
       }
-    } else {
-      for (int i = 0; i < (bytes_received / sizeof(uint64_t)); i += 8) {
-        // this stores as raw binary. stored as little endian.
-        // e.g. to get the same thing as the human readable above,
-        // flip all the bytes in each 512-bit line.
-        for (int q = 0; q < 8; q++) {
-          fwrite(OUTBUF + (i + q), sizeof(uint64_t), 1, this->tracefile);
-        }
+    }
+  } else {
+    for (int i = 0; i < (bytes_received / sizeof(uint64_t)); i += 8) {
+      // this stores as raw binary. stored as little endian.
+      // e.g. to get the same thing as the human readable above,
+      // flip all the bytes in each 512-bit line.
+      for (int q = 0; q < 8; q++) {
+        fwrite(OUTBUF + (i + q), sizeof(uint64_t), 1, tracefile);
       }
     }
   }
-  return bytes_received;
+}
+
+void tracerv_t::write_header(FILE *file) {
+  fputs(this->clock_info.file_header().c_str(), file);
 }
 
 void tracerv_t::tick() {
+  if (!init_ran) {
```

Comment:
This can be removed - the surrounding system should ensure that init is called before tick.